lib/llm_composer/providers_runner.ex

```diff
@@ -0,0 +1,132 @@
+defmodule LlmComposer.ProvidersRunner do
+  @moduledoc """
+  Handles provider execution logic including fallback strategies, routing,
+  and error handling for multiple provider configurations.
+  """
+
+  alias LlmComposer.LlmResponse
+  alias LlmComposer.Message
+  alias LlmComposer.Settings
+
+  require Logger
+
+  @type messages :: [Message.t()]
+
+  @doc """
+  Runs provider execution with fallback support for multiple providers.
+  """
+  @spec run(messages(), Settings.t(), Message.t()) :: {:ok, any()} | {:error, atom()}
+  def run(messages, %Settings{providers: [{provider, provider_opts}]} = settings, system_msg) do
+    provider_opts = get_provider_opts(provider_opts, settings)
+    provider.run(messages, system_msg, provider_opts)
+  end
+
+  def run(messages, %Settings{providers: providers} = settings, system_msg)
+      when is_list(providers) and length(providers) > 1 do
+    router = get_provider_router()
+
+    if Process.whereis(router) == nil do
+      {:error, :provider_router_not_started}
+    else
+      do_run(router, providers, messages, system_msg, settings)
+    end
+  end
+
+  # old case, TODO: remove for llm_composer 0.12.0
+  def run(
+        messages,
+        %Settings{provider: provider, provider_opts: provider_opts} = settings,
+        system_msg
+      )
+      when provider != nil do
+    provider_opts = get_provider_opts(provider_opts, settings)
+    provider.run(messages, system_msg, provider_opts)
+  end
+
+  def run(_messages, %Settings{}, _system_msg) do
+    {:error, :no_providers_configured}
+  end
+
+  @spec do_run(module(), [{module(), keyword()}], messages(), Message.t(), Settings.t()) ::
+          {:ok, any()} | {:error, atom()}
+  defp do_run(router, all_providers, messages, system_msg, settings) do
```

Comment:
I would rename `do_run/4` to something like `run_with_fallback/4`.

When I use something like `run` and `do_run` the former calls the latter.